include/osquery/logger.h

```diff
@@ -299,6 +299,40 @@ Status logQueryLogItem(const QueryLogItem& item, const std::string& receiver);
 Status logSnapshotQuery(const QueryLogItem& item);
 
 /**
+ * @brief Stop forwarding logs to the registered logger plugins
+ *
+ * @return The current forwarding state
+ *
+ * Sometimes, it is useful to turn off log forwarding and force status logs to
+ * be buffered. One example is with handling of rocksdb status logs; if those
+ * status logs are forwarded, it can cause a deadlock inside rocksdb, if the
+ * logger plugin tries to call back into rocksdb.
+ *
+ * This method allows one to halt log forwarding and leave the log sink in a
+ * locked/non-forwarding state. Any log requests in this state are guaranteed
+ * to be buffered. Callback loops can thus be avoided.
+ *
+ * While the sink is in a locked state, no other thread can manipulate the
+ * forwarding state.
+ *
+ * Caller must make a corresponding call to restoreForwardingAndUnlock to
+ * re-enable log forwarding.
+ */
+
+bool haltForwardingAndLock();
+
+/**
+ * @brief Restore forwarding state to the specified state and unlock
+ *
+ * @param The forwarding state to restore to
+ *
+ * Matching call to haltForwardingAndLock. Restores the forwarding state to
+ * that specified, and unlocks the forwarding state.
+ */
+
+void restoreForwardingAndUnlock(bool forward);
```

Comment:
This is scary, it allows callers to potentially "forget" to unlock. Maybe like `LoggerDisabler` this can be implemented as a class (eg `LoggerLocker`) then move the halt into the ctor, and restore into the dtor?